crates/rpc/rpc/src/testing.rs

```diff
@@ -0,0 +1,267 @@
+//! Implementation of the `testing` namespace.
+//!
+//! This exposes `testing_buildBlockV1`, intended for non-production/debug use.
+
+use alloy_consensus::Transaction as ConsensusTransaction;
+use alloy_evm::Evm;
+use alloy_primitives::{B256, U256};
+use alloy_rlp::Decodable;
+use alloy_rpc_types_engine::ExecutionPayloadEnvelopeV4;
+use async_trait::async_trait;
+use jsonrpsee::core::RpcResult;
+use reth_chainspec::ChainSpec;
+use reth_ethereum_engine_primitives::{BuiltPayloadConversionError, EthBuiltPayload};
+use reth_ethereum_primitives::{Block as EthBlock, EthPrimitives, Receipt, TransactionSigned};
+use reth_evm_ethereum::EthEvmConfig;
+use reth_primitives_traits::{Block as BlockTrait, Recovered, SealedBlock, SignerRecoverable};
+use reth_rpc_api::{TestingApiServer, TestingBuildBlockRequestV1};
+use reth_rpc_server_types::result::{internal_rpc_err, invalid_params_rpc_err};
+use std::{marker::PhantomData, sync::Arc};
+use tokio::{runtime::Handle, sync::Semaphore};
+
+// EVM / payload building
+use reth_chainspec::{ChainSpecProvider, EthereumHardforks};
+use reth_ethereum_payload_builder::EthereumBuilderConfig;
+use reth_evm::{execute::BlockBuilder, ConfigureEvm, NextBlockEnvAttributes};
+// BlockExecutionError is unused after inlining error handling; keep import minimal.
+use reth_revm::{database::StateProviderDatabase, db::State};
+use reth_storage_api::{BlockReaderIdExt, StateProviderFactory};
+
+/// Trait for types that can build a block for `testing_buildBlockV1`.
+#[async_trait]
+pub trait TestingBlockBuilder: Send + Sync + std::fmt::Debug + 'static {
+    /// Build a block according to the testing API request.
+    async fn build_block(
+        &self,
+        request: TestingBuildBlockRequestV1,
+    ) -> RpcResult<ExecutionPayloadEnvelopeV4>;
+}
```

Comment:
I think this api could potentially be useful for other networks as well, so we can make an effort here to make this generic over the types, like N: NodePrimitives perhaps, or at least the actual TestingApi